### PR TITLE
Debugger stepping does not understand recursion (bug 38025)

### PR DESCRIPTION
**Pull request by xmcclure**
Testing with the debugger showed multiple failures around recursive
functions:

1. Step-out breakpoints can trigger on inner frames.
**Solution**: Extend existing inner-frame test from step-over to step out
also.

2. "Same source line, continuing" test triggers even on wrong frame
   (affects both step over and step out)
**Solution**: Check frame as part of this test

3. When step breakpoints pick next sequence points, they can register
   multiple breakpoints at the same place if a recursive call happens to
   exit there. (Possibly harmless but may have resulted in duplicate
   debugger events being generated in one test.)
**Solution**: Check for duplicates in step-request breakpoint list before
adding new breakpoints

4. Mono.Debugging.Soft in the debugger-libs repo sometimes resumes
   after breakpoints if it decides Mono.Debugger.Soft has not stepped
   far enough, but the check to trigger this was not recursion-aware.
**Solution**: This one is not addressed in this commit, but a fix was
checked in to debugger-libs. Any code using this commit therefore
needs to be run in conjunction with debugger-libs dfbf6d8b13.

Tests were also added for issues 1-3 above.

### CODE REVIEW

**Comment by xmcclure** [comment]
The functional parts of this commit are in debugger-agent.c, the rest is tests.

I would like to check this in to the 4.5 branch also.

File: mono/mini/debugger-agent.c [hunk 1]
```diff
@@ -5167,8 +5203,10 @@ ss_start (SingleStepReq *ss_req, MonoMethod *method, SeqPoint* sp, MonoSeqPointI
 			for (i = 0; i < sp->next_len; i++) {
 				next_sp = &next[i];
 
-				bp = set_breakpoint (method, next_sp->il_offset, ss_req->req, NULL);
-				ss_req->bps = g_slist_append (ss_req->bps, bp);
+				if (!ss_bp_is_duplicate(ss_req->bps, method, next_sp->il_offset)) {
```
Comment (kumpera) [inline mono/mini/debugger-agent.c]: This introduces quadratic behavior in the debugger, can sizes here be big enough that this would be an issue?

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -5073,6 +5090,25 @@ ss_stop (SingleStepReq *ss_req)
 /*
  * ss_start:
  *
+ * Check if breakpoint would collide with one already in list. Print debug trace before returning if so.
+ */
+static gboolean ss_bp_is_duplicate(GSList *bps, MonoMethod *method, guint32 il_offset)
```
Comment (kumpera) [inline mono/mini/debugger-agent.c]: Formating of the return value.

What about changing the meaning of this function to be the opposite so all checks don't need to negate it? Something like ss_bp_is_unique.

**Comment by kumpera** [comment]
Ignore 4.5, we're doing 4.5.1 soon.

File: mono/mini/debugger-agent.c (hunk 1 above)
Comment (vargaz) [inline mono/mini/debugger-agent.c]: It could happen with the seq points at the beginning of a huge switch statement for example.
We might want to fall back to a hash table for entries > 10 for example.

Comment (xmcclure) [inline mono/mini/debugger-agent.c]: @vargaz, I'm going to go ahead and implement the hash table idea, however this seems to me to add dangerous complexity for an edge case. I would like to add a test case for this situation. Could you give me advice on how to reproduce this? Make a large switch and then break on the switch() itself, I guess?

Comment (xmcclure) [inline mono/mini/debugger-agent.c]: Actually turns out the default tests trigger this.

Comment (vargaz) [inline mono/mini/debugger-agent.c]: Yes, stepping from the switch itself would work, since that seq point has many successors.

**Comment by xmcclure** [comment]
I think that addresses all comments. I made 8 the line for switching to the hash (ie after 8 items are in the list, it builds the hash)

I ran the AArch64 debugger test many times on our aarch64 box, I did not reproduce the crash that the first check saw. If we don't see it on this second run I'm going to assume it's transient.

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs [hunk 3]
```diff
@@ -204,7 +204,7 @@ public static void ss_recursive_chaotic_trap (int n, List<RecursiveChaoticPoint>
 	{
 		// Depth is calculated as:
 		// Main + single_stepping + ss_recursive_chaotic + (n is 12 at outermost frame and 0 at innermost frame) + ss_recursive_chaotic_trap
-		trace.Add (new RecursiveChaoticPoint (true, "ss_recursive_chaotic_trap", 12 - n + 5));
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: The comment still says 12.

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs [hunk 4]
```diff
@@ -214,7 +214,7 @@ public static void ss_recursive_chaotic_at (string at, int n, List<RecursiveChao
 		if (didLast) {
 			// Depth is calculated as:
 			// Main + single_stepping + ss_recursive_chaotic + (n is 12 at outermost frame and 0 at innermost frame)
-			trace.Add (new RecursiveChaoticPoint (false, "ss_recursive_chaotic_" + at, 12 - n + 4));
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: Ditto.

**Comment by vargaz** [comment]
Looks good.

**Comment by xmcclure** [comment]
Thanks Zoltan

**Comment by xmcclure** [comment]
Merging since the current build failures seem transient/unrelated to PR and the only change since the last (CI'd) commit was comments
